### PR TITLE
Add server request/response validation

### PR DESCRIPTION
**Pull request by dcolligan**
Issue #175

### CODE REVIEW

File: ga4gh/backend.py [hunk 1]
```diff
@@ -286,6 +292,18 @@ def startProfile(self):
     def endProfile(self):
         pass
 
+    def validateRequest(self, jsonDict, requestClass):
+        if ('REQUEST_VALIDATION' in flaskapp.app.config and
+                flaskapp.app.config['REQUEST_VALIDATION']):
+            if not requestClass.validate(jsonDict):
+                raise backendExceptions.RequestValidationFailureException
```
Comment (jeromekelleher) [inline ga4gh/backend.py]: missing brackets here? ie, ` raise backendExceptions.RequestValidationFailureException()`

File: ga4gh/backend.py [hunk 2]
```diff
@@ -286,6 +292,18 @@ def startProfile(self):
...
+
+    def validateResponse(self, jsonDict, responseClass):
+        if ('RESPONSE_VALIDATION' in flaskapp.app.config and
+                flaskapp.app.config['RESPONSE_VALIDATION']):
+            if not responseClass.validate(jsonDict):
+                raise backendExceptions.ResponseValidationFailureException
```
Comment (jeromekelleher) [inline ga4gh/backend.py]: and here.

**Comment by jeromekelleher** [comment]
I'm really not keen at all on the backend depending on flask. Is there any other way we can do this? We just need to do some configuration on the backend. Can't we just set some instance variable during configure? i.e.
```python
def configure(...):
    # All the stuff we already have...
    backend = backend.Backend(app.config["DATADIR"])
    backend.setRequestValidation(app.config["REQUEST_VALIDATION"])
    backend.setResponseValidation(app.config["REQUEST_VALIDATION"])
    app.backend = backend
```
This greatly simplifies things, and means we don't have to pass around an ugly config dictionary and do lots of testing on string constants. This will require replumbing things a bit, but solving #186 gets us most of the way there.

File: ga4gh/backend.py (hunk 1 above)
Comment (dcolligan) [inline ga4gh/backend.py]: Exceptions raised like this are implicitly instantiated.  Do we want to add parentheses as a matter of style?

Related: https://www.python.org/dev/peps/pep-0317/

Comment (jeromekelleher) [inline ga4gh/backend.py]: Yeah, I think we should. It's just making an awkward special case out of no-parameter exceptions.

**Comment by dcolligan** [comment]
@jeromekelleher pushed an update

File: ga4gh/backend.py [hunk 3]
```diff
@@ -286,6 +293,22 @@ def startProfile(self):
     def endProfile(self):
         pass
 
+    def validateRequest(self, jsonDict, requestClass):
```
Comment (jeromekelleher) [inline ga4gh/backend.py]: Document these methods please -- all public methods should have docstrings.

File: ga4gh/flaskapp.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
+"""
+Contains the actual flask startup code
+"""
```
Comment (jeromekelleher) [inline ga4gh/flaskapp.py]: I'm not really seeing the benefit of this new module. We still have lots of flasky stuff in frontend.py, so we're not breaking flask from everything else. I'd vote for keeping things are they are here.

**Comment by dcolligan** [comment]
@jeromekelleher pushed a new change addressing your concerns

**Comment by jeromekelleher** [comment]
Excellent, thanks @dcolligan. Why do we need to default to not validating requests are responses for testing? If we're failing validation at the moment, please open an issue on this once this PR has been merged.

+1.

**Comment by dcolligan** [comment]
Because in some cases tests send data to the server that won't validate like `self.sendVariantsSearch('{"variantSetIds": [1, 2]}')` in `test_views.py`.  It's only for a small subset of tests that we default to non-validation.

Yes, I have verified that request-responses work (see, this is why we need an end to end test!).

**Comment by dcolligan** [comment]
(And the fact that only the tests with malformed data failed testifies to this change working.)

**Comment by jeromekelleher** [comment]
Thanks @dcolligan, just checking! (And yes, we _do_ need end-to-end tests, I completely agree).

+1.

**Pull request merged**